lightwood/encoders/time_series/rnn.py

```diff
@@ -5,122 +5,171 @@
         if self._prepared:
             raise Exception('You can only call "prepare_encoder" once for a given encoder.')
 
-        self._encoder = EncoderRNNNumerical(hidden_size=self._encoded_vector_size).to(self.device)
-        optimizer = optim.Adam(self._encoder.parameters(), lr=self._learning_rate)
-        criterion = nn.MSELoss()
-
         self._encoder.train()
         for i in range(self._train_iters):
             average_loss = 0
-            for data_point in priming_data:
-                data_tensor = tensor_from_series(data_point, self.device)
+            data_idx = 0
+
+            while data_idx < len(priming_data):
+
+                # batch building
+                data_points = priming_data[data_idx:min(data_idx + batch_size, len(priming_data))]
+                batch = []
+                for dp in data_points:
+                    data_tensor = tensor_from_series(dp, self.device, n_dims=self._n_dims,
+                                                     pad_value=self._eos, max_len=self._max_ts_length)
+                    batch.append(data_tensor)
+
+                # shape: (batch_size, timesteps, n_dims)
+                batch = torch.cat(batch, dim=0).to(self.device)
+                data_idx += batch_size
+
+                # setup loss and optimizer
+                steps = batch.shape[1]
                 loss = 0
+                self.optimizer.zero_grad()
 
-                optimizer.zero_grad()
+                # encode
                 encoder_hidden = self._encoder.initHidden(self.device)
-                next_tensor = data_tensor[0]
-                for tensor_i in range(len(data_tensor)-1):
+                next_tensor = data_tensor[:, 0, :].unsqueeze(dim=1)  # initial input
+
+                for tensor_i in range(steps - 1):
                     rand = np.random.randint(2)
-                    # teach from forward as well as from known tensor alteratively
+                    # teach from forward as well as from known tensor alternatively
                     if rand == 1:
-                        next_tensor, encoder_hidden = self._encoder.forward(data_tensor[tensor_i] , encoder_hidden)
+                        next_tensor, encoder_hidden = self._encoder.forward(
+                            data_tensor[:, tensor_i, :].unsqueeze(dim=1),
+                            encoder_hidden)
                     else:
                         next_tensor, encoder_hidden = self._encoder.forward(next_tensor.detach(), encoder_hidden)
-                    loss += criterion(next_tensor, data_tensor[tensor_i+1])
 
+                    loss += self.criterion(next_tensor, data_tensor[:, tensor_i + 1, :].unsqueeze(dim=1))
+
+                # decode
+                decoder_hidden = encoder_hidden
+                sos_vector = torch.full((batch.shape[0], 1, batch.shape[2]), self._sos,
+                                        dtype=torch.float32).to(self.device)
+                next_tensor = torch.clone(sos_vector)
```

Comment:
Does this need to be cloned here ? since `next_tensor.detach()` is called in the case where this representation is used and in the next iteration the variable will be reassigned anyway.

If it's just here because you might want to remove the `.detach` in the future and that would cause issues in case this tensor isn't cloned (would it ?) or if it's some other reason, feel free to leave it in.

Reply:
Actaully 99% this isn't needed, since one step above `torch.full` is used and that will create a new tensor anyway.